packages/@lwc/engine-core/src/framework/wiring.ts

```diff
@@ -108,57 +109,74 @@ function createConfigWatcher(
-            },
-            setDisconnectedCallback(disconnectCallback: () => void) {
-                // adds this callback into the disconnect bucket so it gets disconnected from parent
-                // the the element hosting the wire is disconnected
-                ArrayPush.call(wiredDisconnecting, disconnectCallback);
-            },
-        });
-        renderer.dispatchEvent(elm, contextRegistrationEvent);
-    });
+
+    let isConnected = false;
+    let context: ContextValue | undefined;
+    const wiredDisconnecting: Array<() => void> = [];
+
+    return {
+        connect() {
+            isConnected = true;
+
+            // This event is responsible for connecting the host element with another
+            // element in the composed path that is providing contextual data. The provider
+            // must be listening for a special dom event with the name corresponding to the value of
+            // `adapterContextToken`, which will remain secret and internal to this file only to
+            // guarantee that the linkage can be forged.
+            const contextRegistrationEvent = new WireContextRegistrationEvent(adapterContextToken, {
+                setNewContext(newContext: ContextValue) {
+                    // eslint-disable-next-line lwc-internal/no-invalid-todo
+                    // TODO: dev-mode validation of config based on the adapter.contextSchema
+                    if (context !== newContext) {
+                        context = newContext;
+                        // Note: when new context arrives, the config will be recomputed and pushed along side the new
+                        // context, this is to preserve the identity characteristics, config should not have identity
+                        // (ever), while context can have identity
+                        if (isConnected) {
+                            computeConfigAndUpdate();
+                        }
+                    }
+                },
+                setDisconnectedCallback(disconnectCallback: () => void) {
+                    // adds this callback into the disconnect bucket so it gets disconnected from parent
+                    // the the element hosting the wire is disconnected
+                    ArrayPush.call(wiredDisconnecting, disconnectCallback);
+                },
+            });
+            renderer.dispatchEvent(elm, contextRegistrationEvent);
+        },
+        disconnect() {
+            isConnected = false;
+
+            for (let i = 0, n = wiredDisconnecting.length; i < n; i++) {
+                wiredDisconnecting[i]();
+            }
+        },
+        get contextValue() {
+            return context;
+        },
+    };
 }
 
-function createConnector(
-    vm: VM,
-    name: string,
-    wireDef: WireDef
-): {
-    connector: WireAdapter;
-    computeConfigAndUpdate: () => void;
-    resetConfigWatcher: () => void;
-} {
+function createConnector(vm: VM, name: string, wireDef: WireDef): WireConnector {
```

Comment:
The amount of closure passing makes the code hard to read. I would convert the connector, config watcher, and context watcher into 3 different classes and pass the object reference around. This would certainly make the code more readable and reduce the overall memory usage due to the closure creation.